DSpace/dspace-angular
# Fixed caching & same entity type relationship with different left/rightwardtype bugs on edit item relationships

## References
* Fixes last remaining bugs noticed during #3060
* Backport PR #3109

## Description
This PR fixes 2 remaining bugs. A caching issue where the selected status of the relationship popup modal would stay cached even after deleting a relationship. And also an issue where only one side of the same entity type relationship would be shown on the edit item relationship page (for example both `isDepartmentOfDivision` & `isDivisionOfDepartment` should be shown). It also hides the item you are currently on in the search results of the popup modal.

## Instructions for Reviewers
List of changes in this PR:
* When removing a relationship the item you are on is invalidated, but the item on the other side of the relationship wasn't, so this has now been fixed in `EditItemRelationshipsService#submit`
* Every search request in the relationship popup modal stayed cached even when it contained an item which was just invalidated. By using `addDependency` this search request is now linked to every item that is returned in the results, this way when one of the items is invalidated the request is also invalidated.
* The current item won't be displayed in the search results in the popup modal anymore (because making a relationship with yourself doesn't really make sense). This was only visible for same entity relationships
* When you have a relationship with the same entity type on both sides it would only show one side. To fix this a new component was created (`EditRelationshipListWrapperComponent`), which will display both sides of the relationship when a relationship has the same entity type on both sides and has different `leftwardType` & `rightwardType`

**Guidance for how to test or review this PR:**
Caching issues:
- On the edit relationships tab open click copy the uuid of one of the objects with whom you can create a relationship and use this a query to find results (normally only one results should be displayed).
- Select the item and save it
- Open the modal again and search for the same uuid, the checkbox should be selected

Display both side of the relationship:
- Add the following relationship config to your `relationship-types.xml`:
   ```xml
   <type>
        <leftType>OrgUnit</leftType>
        <rightType>OrgUnit</rightType>
        <leftwardType>isDepartmentOfDivision</leftwardType>
        <rightwardType>isDivisionOfDepartment</rightwardType>
        <leftCardinality>
            <min>0</min>
        </leftCardinality>
        <rightCardinality>
            <min>0</min>
        </rightCardinality>
        <copyToLeft>true</copyToLeft>
    </type>
    <type>
        <leftType>OrgUnit</leftType>
        <rightType>OrgUnit</rightType>
        <leftwardType>isOrgUnitOfOrgUnit</leftwardType>
        <rightwardType>isOrgUnitOfOrgUnit</rightwardType>
        <leftCardinality>
            <min>0</min>
        </leftCardinality>
        <rightCardinality>
            <min>0</min>
        </rightCardinality>
        <copyToLeft>true</copyToLeft>
    </type>
   ```
- Create the following metadata fields in `relationship-formats.xml` (can also be done though the UI):
   ```xml
    <dc-type>
        <schema>relation</schema>
        <element>isDepartmentOfDivision</element>
    </dc-type>
    <dc-type>
        <schema>relation</schema>
        <element>isDivisionOfDepartment</element>
    </dc-type>
    <dc-type>
        <schema>relation</schema>
        <element>isOrgUnitOfOrgUnit</element>
    </dc-type>
   ```
- Verify that there are 2 separate sections for the first relationship type: a `isDepartmentOfDivision` section & `isDivisionOfDepartment` section
- Verify that for the `isOrgUnitOfOrgUnit` only one section is shown
- Verify that you can't see the item you are on anymore in the list of suggested items

## Checklist
